native/src/main/java/io/ballerina/stdlib/nats/Utils.java

```diff
@@ -126,17 +128,35 @@ public static Object getValueWithIntendedType(Type type, byte[] value) throws BE
                     return ValueCreator.createArrayValue(value);
                 case RECORD_TYPE_TAG:
                     return CloneWithType.convert(type, JsonUtils.parse(strValue));
+                case UNION_TAG:
+                    if (hasStringType((UnionType) type)) {
+                        return StringUtils.fromString(strValue);
+                    }
+                    return getValueFromJson(type, strValue);
```

Comment:
Don't we need to fall through? To address scenarios union types like `byte[]|string payload`? Not sure whether this is a valid use case in messaging :)

Reply:
Its covered using `getValueFromJson(type, strValue);`